pylhc/autosix.py

```diff
@@ -0,0 +1,525 @@
+            mask=Path('my_madx.mask'), # can contain any of the parameters used in replace_dict
+            python=Path('/afs/cern.ch/work/u/user/my_venv/bin/python'),
+            ignore_twissfail_check=False,  # if script prints 'check if twiss failed' or similar
+            replace_dict=dict(
+                # Part of the sixdesk-environment:
+                TURNS=100000,
+                AMPMIN=4, AMPMAX=30, AMPSTEP=2,
+                ANGLES=11,
+                # Examples for mask:
+                BEAM=[1, 4],
+                TUNE=[62.29, 62.30, 62.31, 62.31]
+                OUTPUT='/afs/cern.ch/work/u/user/study_output/',
+                SEED='%SEEDRAN',  # puts '%SEEDRAN' in the mask and lets sixdesk handle this loop
+            ),
+            jobid_mask="B%(BEAM)d-QX%(TUNE)s",
+            # unlock=True,
+            # resubmit=True,
+            # ssh='lxplus.cern.ch',
+        )
+
+
+Upon running the script the job-matrix is created
+(see __Jobs.tfs__ in working directory) and the following stages are run per job:
+
+- ``create_jobs``: create workspace; fill sysenv, sixdeskenv, mask; initialize workspace.
+- ``submit_mask``: submit mask-job to HTCondor. (__interrupt__)
+- ``check_input``: check if sixdesk input is complete.
+- ``submit_sixtrack``: submit sixdesk-jobs to HTCondor. (__interrupt__)
+- ``check_sixtrack_output``: check if all sixdesk jobs are completed.
+- ``sixdb_load``: crate database and load jobs output.
+- ``sixdb_cmd``: calculated DA from database data.
+- ``post_process``: extract data from database, write into _.tfs_ and plot.
+- ``final``: announce everything has finihed
+
+
+To keep track of the stages, they are written into the __stages\_completed.txt__
+in the __autosix\_output__ directory in the workspaces.
+Stages that are written in this file are assumed to be done and will be skipped.
+To rerun a stage, delete the stage and all following stages in that file and
+start your script anew.
+
+The stages are run independently of each job, meaning different jobs can be
+at different stages. E.g if one job has all data for the ``six_db`` analysis
+already, but the others are still running on sixdesk the
+``check_sixtrack_output`` stage will fail for these jobs but the other one will
+just continue.
+
+Because the stages after ``submit_mask`` and ``submit_sixtrack`` need only be
+run after the jobs on HTCondor are completed, these two stages interrupt the
+execution of stages if they have successfully finished. Check your scheduler
+via ``condor_q`` and run your script again after everything is done, to
+have autosix continue its work.
+
+
+For the creation of polar plots, the function
+:func:`pylhc.sixdesk_tools.post_process_da.plot_polar` is available, which is
+used for the basic polar plotting in the ``post_process`` stage, but provides
+more customization features if called manually.
+
+
+Arguments:
+
+*--Required--*
+
+- **mask** *(PathOrStr)*:
+
+    Program mask to use
```

Comment:
madx mask or cpymask?

Reply:
whatever you give as binary, path to madx or a python with cpymad installed

Reply:
or whatever you want to run to produce the sixtrack input... could be a bash file that runs madx...